Cards NavModel

zsoltk opened the pull request:
## Description

* Creates a `:samples:common` module to share images / colors / resources / etc.
* Adds profile images
* Extracts common profile-related classes and composables
* Adds `Cards` NavModel to `navmodel-samples`
* Makes both `Cards` and `AdvancedSpotlight` use the common profile-related images and classes
* Adds a `CardsExampleNode` demo to `:samples:sandbox`
* Adds a `CardsExampleNode` demo to `:samples:app`

## Check list

- [x] I have updated `CHANGELOG.md` if required.
- [x] I have updated documentation if required.

zsoltk committed:
Add :samples:common with profile images

zsoltk committed:
Add Profile class

zsoltk committed:
Use new profiles in spotlight advanced

zsoltk committed:
Move ProfileCard to common

zsoltk committed:
Add cards navmodel

zsoltk committed:
Add cards example to sandbox

zsoltk committed:
Move NavModel examples to top in sandbox

KovalevAndrey commented on navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/transitionhandler/CardsTransitionHandler.kt:
```diff
@@ -0,0 +1,160 @@
+        // If we only rotated the cards at the indicate vote states,
+        // they'd just rotate around their own center.
+        // We'll also want to offset the cards on an arc for a nice visual effect.
+        val angularOffset = derivedStateOf {
```
remember `derivedStateOf`

KovalevAndrey commented on navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/CardsOnScreenResolver.kt:
```diff
@@ -0,0 +1,18 @@
+            is State.Bottom,
+            is State.Top -> true
+            is State.Queued,
+            is State.IndicateLike,
```
shouldn't indicate like/pass be onScreen?

KovalevAndrey commented on sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/cardsexample/CardsExampleNode.kt:
```diff
@@ -0,0 +1,86 @@
+                .fillMaxSize()
+                .padding(padding),
+            navModel = cards,
+            transitionHandler = rememberCardsTransitionHandler { spring(stiffness = Spring.StiffnessVeryLow) }
```
it's `Spring.StiffnessVeryLow` by default

KovalevAndrey commented on samples/common/proguard-rules.pro:
```diff
@@ -0,0 +1,21 @@
+# Add project specific ProGuard rules here.
```
delete?

KovalevAndrey commented on navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/operation/CardsOperation.kt:
```diff
@@ -0,0 +1,6 @@
+import com.bumble.appyx.core.navigation.Operation
+import com.bumble.appyx.navmodel.cards.Cards
+
+sealed interface CardsOperation<T> : Operation<T, Cards.State>
```
sealed? I think we agreed on removing it and allowing clients define custom operations

KovalevAndrey commented on navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/operation/TopCardOperation.kt:
```diff
@@ -0,0 +1,19 @@
+    private val newTargetState: Cards.State
+    ) : CardsOperation<T> {
+
+    override fun isApplicable(elements: CardsElements<T>): Boolean =
+        true
```
```suggestion
    override fun isApplicable(elements: CardsElements<T>): Boolean =
        elements.isNotEmpty()
```

zsoltk committed:
Fix IndicateLike & IndicatePass on screen resolution

zsoltk committed:
Remember derivedStateOf

zsoltk committed:
Remove redundant params

zsoltk committed:
Tweak layout

zsoltk committed:
Open CardsOperation

zsoltk committed:
Add proper condition to TopCardOperation.isApplicable

zsoltk committed:
Add lint baseline

zsoltk committed:
Cleanup dependencies

zsoltk committed:
Merge branch 'main' into cards-navmodel

zsoltk committed:
Remove themes.xml from samples

zsoltk committed:
Remove coil dependency from sandbox

LachlanMcKee commented on samples/common/build.gradle.kts:
```diff
@@ -0,0 +1,43 @@
+}
+
+android {
+    namespace = "com.bumble.appyx.common"
```
`com.bumble.appyx.sample.common` - otherwise this could get confusing in the future if we ever create another common module.

LachlanMcKee commented on samples/common/build.gradle.kts:
```diff
@@ -0,0 +1,43 @@
+    id("com.android.library")
+    id("kotlin-android")
+    id("kotlin-parcelize")
+    id("appyx-publish-android")
```
We don't want to publish this I assume? you should remove this, otherwise we will accidentally publish it next release.

zsoltk replied:
Good catch!! Copy-paste remnant.

LachlanMcKee commented on navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/transitionhandler/CardsTransitionHandler.kt:
```diff
@@ -0,0 +1,160 @@
+    private val transitionSpec: TransitionSpec<Cards.State, Float> = { spring(stiffness = Spring.StiffnessVeryLow) }
+) : ModifierTransitionHandler<T, Cards.State>() {
+
+    data class Props(
```
This can be private

LachlanMcKee reviewed (commented):
Goods great! My only primary concern (other than the comments):

Are these photos definitely fake? i.e. deep fakes? worried about the legality of having these in the repository otherwise. Perhaps we should add a README somewhere that mentions this?

LachlanMcKee reviewed (changes requested):
The publishing issue is definitely a blocker (forgot to request changes)

zsoltk committed:
Remove accidental publish from :samples:common

zsoltk committed:
Make Props private

zsoltk committed:
Change .common to .samples.common

zsoltk committed:
Merge branch '1.x' into cards-navmodel

zsoltk committed:
Merge branch '1.x' into cards-navmodel

zsoltk committed:
Replace images with halloween-themed ones

zsoltk committed:
Add dating cards example to sample app

zsoltk committed:
Center buttons horizontally

zsoltk committed:
Update CHANGELOG.md

zsoltk committed:
Rename <T> to <NavTarget>

zsoltk committed:
Optimise imports

zsoltk committed:
Add documentation

zsoltk committed:
Replace image on landing page

zsoltk committed:
Merge branch '1.x' into cards-navmodel

Conflicts:
	CHANGELOG.md
	samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/container/ContainerNode.kt

zsoltk committed:
Fix cards example inclusion in sandbox after merge

zsoltk commented:
@LachlanMcKee ready to review. Summary of recent changes:

- Replaced images with halloween-themed ones
- Added cards example to both sandbox and main sample app
- Added a documentation page

KovalevAndrey commented on samples/navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/transitionhandler/CardsTransitionHandler.kt:
```diff
@@ -0,0 +1,160 @@
+            transitionSpec = transitionSpec,
+            targetValueByState = { it.toProps().positionalOffsetX.value })
+
+        val zIndex = transition.animateFloat(
```
why do we need to animate zIndex? updating the value should suffice?

KovalevAndrey commented on samples/navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/transitionhandler/CardsTransitionHandler.kt:
```diff
@@ -0,0 +1,160 @@
+        // If we only rotated the cards at the indicate vote states,
+        // they'd just rotate around their own center.
+        // We'll also want to offset the cards on an arc for a nice visual effect.
+        val angularOffset = remember { derivedStateOf {
```
we could use delegated properties here and down below

KovalevAndrey commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/spotlightadvancedexample/SpotlightAdvancedExampleNode.kt:
```diff
@@ -55,52 +55,17 @@ class SpotlightAdvancedExampleNode(
-            buildContext = buildContext
-        )
+        when (navTarget) {
+            is NavTarget.ProfileCard -> node(buildContext) { ProfileCard(navTarget.profile) }
```
we could resolve to `ProfileCardNode` instead

zsoltk replied:
we could also do the opposite and remove `ProfileCardNode` altogether, since it kinda adds no value at this point

KovalevAndrey commented on samples/common/build.gradle.kts:
```diff
@@ -0,0 +1,42 @@
+        sourceCompatibility = JavaVersion.VERSION_1_8
+        targetCompatibility = JavaVersion.VERSION_1_8
+    }
+    kotlinOptions {
```
kotlinOptions and compileOptions are redundant here

LachlanMcKee commented on samples/common/build.gradle.kts:
```diff
@@ -0,0 +1,42 @@
+
+        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
+    }
+    compileOptions {
```
We also don't specify `compileOptions` in `build.gradle.kts` either

LachlanMcKee commented on samples/navmodel-samples/lint-baseline.xml:
```diff
@@ -1,4 +1,15 @@
-<issues format="6" by="lint 7.1.3" type="baseline" client="gradle" dependencies="false" name="AGP (7.1.3)" variant="all" version="7.1.3">
+<issues format="6" by="lint 7.2.2" type="baseline" client="gradle" dependencies="false" name="AGP (7.2.2)" variant="all" version="7.2.2">
+
+    <issue
```
It looks like we are suppressing this in code instead now (based on all other usages of this supression): `@SuppressLint("ModifierFactoryExtensionFunction")`

LachlanMcKee commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/cardsexample/CardsExampleNode.kt:
```diff
@@ -0,0 +1,84 @@
+
+    @Composable
+    override fun View(modifier: Modifier) {
+        val padding = remember { 20.dp }
```
It seems strange to remember this. I've read that Google actually suggests reusing the modifier. Maybe that's an even better optimisation? (though it does seem a tad ugly)

https://developer.android.com/jetpack/compose/modifiers#best_practices_for_reusing_modifiers

zsoltk replied:
I don't even know how that got there :D Probs was accidental.

LachlanMcKee commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/cardsexample/CardsExampleNode.kt:
```diff
@@ -0,0 +1,84 @@
+    override fun View(modifier: Modifier) {
+        val padding = remember { 20.dp }
+        Children(
+            modifier = Modifier
```
use the passed in modifier here rather than creating a new one? (this could potentially contradict my comment above)

`Modifer` to `modifier`

LachlanMcKee commented on samples/sandbox/src/main/kotlin/com/bumble/appyx/sandbox/client/spotlightadvancedexample/SpotlightAdvancedExampleNode.kt:
```diff
@@ -33,20 +37,16 @@ import com.bumble.appyx.navmodel.spotlightadvanced.operation.previous
-        ),
+        items = Profile.allProfiles
+            .shuffled()
+            .take(listOf(5, 7, 9, 11).random())
```
This could easily break if the size of `allProfiles` becomes smaller. I guess it's fine for now, but maybe a safer solution would be better long term

LachlanMcKee commented on samples/app/src/main/kotlin/com/bumble/appyx/app/node/samples/SamplesContainerNode.kt:
```diff
@@ -123,9 +128,14 @@ class SamplesContainerNode(
+        ) {
             Button(onClick = { backStack.replace(NavTarget.OnboardingScreen) }) {
-                Text("Onboarding")
+                Text("What's Appyx?")
```
I think `What is Appyx?` sounds more professional

LachlanMcKee commented on samples/app/src/main/kotlin/com/bumble/appyx/app/node/samples/CardsExampleNode.kt:
```diff
@@ -0,0 +1,84 @@
+    override fun View(modifier: Modifier) {
+        val padding = remember { 20.dp }
+        Children(
+            modifier = Modifier
```
Use the `modifier` variable instead of `Modifier`

LachlanMcKee commented on samples/common/src/main/java/com/bumble/appyx/samples/common/profile/Profile.kt:
```diff
@@ -0,0 +1,96 @@
+
+@Parcelize
+@Suppress("MagicNumber")
+data class Profile(
```
`Stable`/`Immutable`? Have you run the Compose compiler metrics on this change? worried it might introduce a non-skippable Composable.

LachlanMcKee commented on samples/common/src/main/java/com/bumble/appyx/samples/common/profile/Profile.kt:
```diff
@@ -0,0 +1,96 @@
+    val age: Int = Random.nextInt(20, 35),
+    val city: String = "London",
+    val uni: String = "Oxford University",
+    @DrawableRes val drawable: Int = allProfiles.random().drawable
```
rename to `drawableRes` or `drawableId`? Looking elsewhere in the code I thought this data class was keeping the actual image in memory.

LachlanMcKee commented on samples/common/src/main/java/com/bumble/appyx/samples/common/profile/ProfileCardNode.kt:
```diff
@@ -24,48 +24,44 @@ import com.bumble.appyx.core.node.Node
+                    .padding(start = 24.dp, bottom = 36.dp)
             ) {
-                Text(text = "$name, $age", color = Color.White, fontSize = 22.sp)
+                Text(text = "${profile.name}, ${profile.age}", color = Color.White, fontSize = 30.sp)
```
Perhaps we could use TextStyles instead of specifying the size? `h1`/`p1` etc in the app in the future? It would make the design language more consistent.

LachlanMcKee reviewed (commented):
Many of the images in this PR use a different size/aspect ratio. Perhaps it would be good to adjust these before merging? (I know we can crop in code, however it would be more efficient to not load unnecessary pixels)

zsoltk committed:
Use ProfileCardNode

zsoltk committed:
Use delegation

zsoltk committed:
Remove redundant config

zsoltk committed:
Use @SuppressLint instead of baseline

zsoltk committed:
Remove unnecessary remember

zsoltk committed:
Use passed in Modifier instead of a new one

zsoltk committed:
Reword text

zsoltk committed:
Add @Immutable on Profile

zsoltk committed:
Rename drawable -> drawableRes

zsoltk committed:
Skip lint on images

KovalevAndrey commented on samples/navmodel-samples/src/main/kotlin/com/bumble/appyx/navmodel/cards/transitionhandler/CardsTransitionHandler.kt:
```diff
@@ -0,0 +1,162 @@
+            Offset(x.toFloat(), y.toFloat())
+        }}
+
+        val rotationZ = transition.animateFloat(
```
let's use delegation everywhere?

zsoltk committed:
Use delegation
